extension/intents/nicknames/routineExecutor.js

```diff
@@ -0,0 +1,102 @@
+/* globals log */
+
+import * as intentRunner from "../../background/intentRunner.js";
+
+export class RoutineExecutor {
+  constructor(
+    routineName,
+    subcommands,
+    index = 0,
+    states = [{}],
+    forIndex = 0
+  ) {
+    this.routineName = routineName;
+    this.subcommands = subcommands;
+    this.index = index;
+    this.states = states;
+    this.forIndex = forIndex;
+  }
+
+  mapState(subcommand, state) {
+    const copySubcommand = JSON.parse(JSON.stringify(subcommand));
+    for (const slot in subcommand.slots) {
+      if (subcommand.slots[slot] in state) {
+        copySubcommand.slots[slot] = state[subcommand.slots[slot]];
+      }
+    }
+    return copySubcommand;
+  }
+
+  async runSubcommand(subcommand) {
+    let hadError = false;
+    let errorMessage = null;
+
+    log.info(
+      "  Running subintent",
+      subcommand,
+      subcommand.name,
+      subcommand.slots,
+      this.index
+    );
```

Comment:
The log message should probably go after `this.mapState()` so it shows the substitution.